sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/VertxAsyncHttpClientBuilder.java

```diff
@@ -0,0 +1,306 @@
+                InetSocketAddress proxyAddress = buildProxyOptions.getAddress();
+
+                if (proxyAddress != null) {
+                    vertxProxyOptions.setHost(proxyAddress.getHostName());
+                    vertxProxyOptions.setPort(proxyAddress.getPort());
+                }
+
+                String proxyUsername = buildProxyOptions.getUsername();
+                String proxyPassword = buildProxyOptions.getPassword();
+                if (!CoreUtils.isNullOrEmpty(proxyUsername) && !CoreUtils.isNullOrEmpty(proxyPassword)) {
+                    vertxProxyOptions.setUsername(proxyUsername);
+                    vertxProxyOptions.setPassword(proxyPassword);
+                }
+
+                ProxyOptions.Type type = buildProxyOptions.getType();
+                if (type != null) {
+                    try {
+                        ProxyType proxyType = ProxyType.valueOf(type.name());
+                        vertxProxyOptions.setType(proxyType);
+                    } catch (IllegalArgumentException e) {
+                        throw LOGGER.logExceptionAsError(new IllegalArgumentException("Unknown Vert.x proxy type: " + type.name(), e));
+                    }
+                }
+
+                String nonProxyHosts = buildProxyOptions.getNonProxyHosts();
+                if (!CoreUtils.isNullOrEmpty(nonProxyHosts)) {
+                    for (String nonProxyHost : desanitizedNonProxyHosts(nonProxyHosts)) {
+                        this.httpClientOptions.addNonProxyHost(nonProxyHost);
+                    }
+                }
+
+                this.httpClientOptions.setProxyOptions(vertxProxyOptions);
+            }
+        }
+
+        if (shutdownHookRequired) {
+            Runtime.getRuntime().addShutdownHook(new Thread(getVertxCloseRunnable()));
+        }
+
+        io.vertx.core.http.HttpClient client = this.vertx.createHttpClient(this.httpClientOptions);
+        return new VertxAsyncHttpClient(client, this.vertx);
+    }
+
+    /**
+     * Reverses non proxy host string sanitization applied by {@link ProxyOptions}.
+     *
+     * This is necessary as Vert.x will apply its own sanitization logic.
+     *
+     * @param nonProxyHosts The list of non proxy hosts
+     * @return String array of desanitized proxy host strings
+     */
+    private String[] desanitizedNonProxyHosts(String nonProxyHosts) {
+        String desanitzedNonProxyHosts = NON_PROXY_HOST_DESANITIZE.matcher(nonProxyHosts)
+            .replaceAll("");
+
+        desanitzedNonProxyHosts = NON_PROXY_HOST_DOT_STAR.matcher(desanitzedNonProxyHosts)
+            .replaceAll("*");
+
+        return NON_PROXY_HOSTS_SPLIT.split(desanitzedNonProxyHosts);
+    }
+
+    /**
+     * Gets a {@link Runnable} to close the embedded {@link Vertx} instance on shutdown.
+     *
+     * @return The {@link Runnable} action to close the embedded {@link Vertx} instance
+     */
+    private Runnable getVertxCloseRunnable() {
```

Comment:
It would be better to pass the Vertx client that needs to be closed here, if a reference to the builder is maintained and the Vertx variable is changed this may leak a Vertx HttpClient to close

Reply:
Hopefully I've interpreted your comment correctly. I modified things so that the local `Vertx` variable (see previous change above) is passed to the shutdown hook.

Reply:
Yup! You interpreted exactly what I meant